### ISSUE TITLE
false positive UndefinedClass for trait ::class

### ISSUE THREAD

**Issue by ostrolucky** [vimeo/psalm#714]
```php
trait A {
  public function foo(): string {
    return B::class;
  }
}

trait B {
  
}

class C {
  use A;
}
```

> ERROR: UndefinedClass - 5:12 - Class or interface B does not exist

https://getpsalm.org/r/a5d543dbbd

**Comment by muglug** [comment]
`B` isn't a class, though...

**Comment by ostrolucky** [comment]
Interface isn't a class either. This is just poor naming from PHP side. Nevertheless it is commonly used for all of them.

https://3v4l.org/bKar3

**Comment by muglug** [comment]
Oh, it works, but since traits don't exist as compile-time entities (interfaces do) the syntax is meaningless. `NonExistentClass::class` also works - PHP doesn't do any checks for `::class` constants.

**Comment by ostrolucky** [comment]
Yes and that's good, otherwise strings would have to be used for checking if classes exist.

So, what's the conclusion? I think this is important to support.

**Comment by muglug** [comment]
> I think this is important to support.

Why?

**Comment by ostrolucky** [comment]
Ah just because it's common imho.

**Comment by muglug** [comment]
I just don't see where you would use it in code

**Comment by ostrolucky** [comment]
Here is one https://github.com/Elao/PhpEnums/blob/master/src/AutoDiscoveredValuesTrait.php
